### PR TITLE
Allow updating of `certificate_arn` argument in `aws_cognito_user_pool_domain` resource

### PR DESCRIPTION
**Pull request by austinvalle**
- This PR adds the ability to update the `aws_cognito_user_pool_domain` resource, specifically only the `certificate_arn` argument.
- The `certificate_arn` can only be updated, it cannot be added or removed or it will force a re-create of the resource.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #14733

Output from acceptance testing:
```
 $ make testacc TESTS=TestAccCognitoIDPUserPoolDomain PKG=cognitoidp
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/cognitoidp/... -v -count 1 -parallel 20 -run='TestAccCognitoIDPUserPoolDomain'  -timeout 180m
=== RUN   TestAccCognitoIDPUserPoolDomain_basic
=== PAUSE TestAccCognitoIDPUserPoolDomain_basic
=== RUN   TestAccCognitoIDPUserPoolDomain_custom
=== PAUSE TestAccCognitoIDPUserPoolDomain_custom
=== RUN   TestAccCognitoIDPUserPoolDomain_customCertUpdate
=== PAUSE TestAccCognitoIDPUserPoolDomain_customCertUpdate
=== RUN   TestAccCognitoIDPUserPoolDomain_disappears
=== PAUSE TestAccCognitoIDPUserPoolDomain_disappears
=== CONT  TestAccCognitoIDPUserPoolDomain_basic
=== CONT  TestAccCognitoIDPUserPoolDomain_customCertUpdate
=== CONT  TestAccCognitoIDPUserPoolDomain_custom
=== CONT  TestAccCognitoIDPUserPoolDomain_disappears
--- PASS: TestAccCognitoIDPUserPoolDomain_disappears (30.30s)
--- PASS: TestAccCognitoIDPUserPoolDomain_basic (31.42s)
--- PASS: TestAccCognitoIDPUserPoolDomain_custom (676.16s)
--- PASS: TestAccCognitoIDPUserPoolDomain_customCertUpdate (1584.20s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/cognitoidp	1586.233s
...
```

### CODE REVIEW

File: internal/service/cognitoidp/user_pool_domain.go [hunk 1]
```diff
@@ -60,6 +62,10 @@ func ResourceUserPoolDomain() *schema.Resource {
 				Computed: true,
 			},
 		},
+		CustomizeDiff: customdiff.ForceNewIfChange("certificate_arn", func(_ context.Context, old, new, meta interface{}) bool {
+			// If the cert arn is being changed to a new arn, don't force new
+			return !(old.(string) != "" && new.(string) != "")
```
Comment (jpm63) [inline internal/service/cognitoidp/user_pool_domain.go:67]: Is there *any* risk that `old` and `new` are not strings, forcing the type assertion to panic?

File: internal/service/cognitoidp/user_pool_domain.go [hunk 2]
```diff
@@ -142,6 +148,38 @@ func resourceUserPoolDomainRead(d *schema.ResourceData, meta interface{}) error
 	return nil
 }
 
+func resourceUserPoolDomainUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*conns.AWSClient).CognitoIDPConn
```
Comment (jpm63) [inline internal/service/cognitoidp/user_pool_domain.go:152]: Simiar question to above - is there any risk `meta` is not `*conns.AWSClient`?

File: internal/service/cognitoidp/wait.go [hunk 3]
```diff
@@ -54,3 +54,24 @@ func waitUserPoolDomainCreated(conn *cognitoidentityprovider.CognitoIdentityProv
...
+			cognitoidentityprovider.DomainStatusTypeUpdating,
+		},
+		Target: []string{
+			cognitoidentityprovider.DomainStatusTypeActive,
+		},
+		Refresh: statusUserPoolDomain(conn, domain),
+		Timeout: timeout,
+	}
+
+	outputRaw, err := stateConf.WaitForState()
```
Comment (jpm63) [inline internal/service/cognitoidp/wait.go:70]: Should add a guard clause incase `err != nil` and `outputRaw` is somehow malformed

File: internal/service/cognitoidp/user_pool_domain.go (hunk 1 above)
Comment (austinvalle) [reply internal/service/cognitoidp/user_pool_domain.go:67]: So that is totally possible (if the schema was incorrect, or a farther down the line bug), due to the terraform plugin SDK being unable to guarantee the type at compile time.

This function is considered "best effort" according to their documentation [here](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2@v2.17.0/helper/customdiff#ForceNewIfChange), so the error is not expected to be handled and I don't believe I can add more information to the error that would come from a type coercion failure.

I was following some examples of other services and they all seem to be throwing that problem to the wind! lol

File: internal/service/cognitoidp/user_pool_domain.go (hunk 2 above)
Comment (austinvalle) [reply internal/service/cognitoidp/user_pool_domain.go:152]: This `meta` argument is controlled by the AWS provider setup, but again, [not guaranteed](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2@v2.17.0/helper/schema#Resource) at compile-time unfortunately. From looking at other resources in the AWS provider it seems to be common to just type coerce without a net, I decided to follow convention for consistency and let Hashi decide if they want it to be handled differently.

File: internal/service/cognitoidp/user_pool_domain.go [hunk 4]
```diff
@@ -142,6 +148,38 @@ func resourceUserPoolDomainRead(d *schema.ResourceData, meta interface{}) error
...
+
+	domain := d.Get("domain").(string)
```
[the remark this answers is not in the document]
Comment (austinvalle) [reply internal/service/cognitoidp/user_pool_domain.go:154]: Same as others, not guaranteed at compile time via the Terraform plugin SDK, convention for consistency 😵‍💫

File: internal/service/cognitoidp/user_pool_domain.go [hunk 5]
```diff
@@ -142,6 +148,38 @@ func resourceUserPoolDomainRead(d *schema.ResourceData, meta interface{}) error
...
+			CertificateArn: aws.String(v.(string)),
+		}
+		params.CustomDomainConfig = customDomainConfig
+	}
+
+	log.Printf("[DEBUG] Updating Cognito User Pool Domain: %s", params)
+
+	_, err := conn.UpdateUserPoolDomain(params)
+	if err != nil {
+		return fmt.Errorf("Error updating User Pool Domain (%s): %w", d.Id(), err)
```
[the remark this answers is not in the document]
Comment (austinvalle) [reply internal/service/cognitoidp/user_pool_domain.go:173]: This one I'm gonna let them decide how to handle, it's consistent to this specific resource, but looks to be against their current documentation: https://github.com/hashicorp/terraform-provider-aws/blob/main/docs/contributing/error-handling.md#update-error-message-context

I'm leaving it because I didn't want to make changes to the other resources unless they ask for it lol 😉

File: internal/service/cognitoidp/wait.go (hunk 3 above)
Comment (austinvalle) [reply internal/service/cognitoidp/wait.go:74]: Good call, added.

File: internal/service/cognitoidp/user_pool_domain.go (hunk 1 above)
Comment (jpm63) [reply internal/service/cognitoidp/user_pool_domain.go:67]: Okay...in that case, I'd recommend the following. This pattern would also hold for any other comment I had of this nature:

```Go
o, ok := old.(string)
if !ok {
  return fmt.Errorf("some friendly error message that's not a panic and a stack trace")
}

n, ok := new.(string)
if !ok {
  // handle error
}

return !(o != "" && n != "")
```

Any time you do a type assertion in Go, you can assign to one or two vals. If you assign to one, there's no guarantee the type assertion will succeed, and a failure will be a panic. If you assign to two, the second variable is always a `bool` that will let you know if the assertion was successful.

Comment (jpm63) [reply internal/service/cognitoidp/user_pool_domain.go:67]: Strange that `panic` is the preferred error handling mechanism, but oh well...if that's what's used elsewhere, then that's the expectation

File: internal/service/cognitoidp/user_pool_domain.go (hunk 2 above)
Comment (jpm63) [reply internal/service/cognitoidp/user_pool_domain.go:152]: Strange codebase, but if that's the standard, who am I to argue

**Review by jpm63 — APPROVED** [review]

**Comment by jpm63** [comment]
My approval is based on Go syntax, coding style, etc. As to whether this is an appropriate contribution from a business perspective, it feels good to me, but that decision ultimately resides with the Cloud domain

File: internal/service/cognitoidp/user_pool_domain.go (hunk 1 above)
Comment (cmorton9249) [inline internal/service/cognitoidp/user_pool_domain.go:67]: I'm not super familar with the lifecycle of the provider so I'm just gonna put it out there for discussion:

Should the comparison be against the ARN?  I wouldn't expect the ARN to change unless the apply was already done.  If that is the case, where does this come into play?  State update?

If not--what am I missing here?

File: internal/service/cognitoidp/user_pool_domain.go [hunk 6]
```diff
@@ -142,6 +148,38 @@ func resourceUserPoolDomainRead(d *schema.ResourceData, meta interface{}) error
...
 
+func resourceUserPoolDomainUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*conns.AWSClient).CognitoIDPConn
+
+	domain := d.Get("domain").(string)
+	timeout := 60 * time.Minute // Update is only for cert arns on custom domains, which take more time to become active
+
+	params := &cognitoidentityprovider.UpdateUserPoolDomainInput{
+		Domain:     aws.String(domain),
+		UserPoolId: aws.String(d.Get("user_pool_id").(string)),
```
Comment (cmorton9249) [inline internal/service/cognitoidp/user_pool_domain.go:159]: Help me understand all the casting here.  domain was cast as a string on 154.  The result of user_pool_id was cast in the same expression.  This feels completely redundant.

File: internal/service/cognitoidp/user_pool_domain.go [hunk 7]
```diff
@@ -142,6 +148,38 @@ func resourceUserPoolDomainRead(d *schema.ResourceData, meta interface{}) error
...
+
+	log.Printf("[DEBUG] Updating Cognito User Pool Domain: %s", params)
+
+	_, err := conn.UpdateUserPoolDomain(params)
+	if err != nil {
+		return fmt.Errorf("Error updating User Pool Domain (%s): %w", d.Id(), err)
+	}
+
+	if _, err := waitUserPoolDomainUpdated(conn, d.Id(), timeout); err != nil {
+		return fmt.Errorf("error waiting for User Pool Domain (%s) update: %w", d.Id(), err)
```
Comment (cmorton9249) [inline internal/service/cognitoidp/user_pool_domain.go:177]: Consistent error message formatting.  Either it is an Error (173) or an error (177).

File: internal/service/cognitoidp/user_pool_domain.go (hunk 6 above)
Comment (cmorton9249) [reply internal/service/cognitoidp/user_pool_domain.go:159]: Thanks @jpm63 for the insightful explanation!  Turns out I have gotten a little spoiled by reference handling in .NET!  lol

File: internal/service/cognitoidp/user_pool_domain.go [hunk 8]
```diff
@@ -142,6 +148,38 @@ func resourceUserPoolDomainRead(d *schema.ResourceData, meta interface{}) error
 	return nil
 }
 
+func resourceUserPoolDomainUpdate(d *schema.ResourceData, meta interface{}) error {
```
[the remark this answers is not in the document]
Comment (austinvalle) [reply internal/service/cognitoidp/user_pool_domain.go:151]: I agree, however I'm leaving it as it matches the rest of their codebase in terms of consistency: https://github.com/hashicorp/terraform-provider-aws/blob/06ebaf7182bf783e334d3c479d4efe2dd8377401/internal/service/s3/bucket.go#L701

File: internal/service/cognitoidp/user_pool_domain.go (hunk 1 above)
Comment (austinvalle) [reply internal/service/cognitoidp/user_pool_domain.go:67]: Good question, so this function is called during a `terraform plan`, and the output is used to determine if the resource must be deleted then re-created, or if the resource can be updated in-place.

In this use-case, user has updated the cert ARN in their terraform to a new cert ARN, which is reflected here during the plan. This code is executed and asserts that an old value existed (previous cert arn) and a new value is being put in its place. (new cert arn).

TLDR: If terraform is not replacing an old cert ARN with a new cert ARN, then we want to force a replacement of the resource, this function allows us to enforce that.

Comment (cmorton9249) [reply internal/service/cognitoidp/user_pool_domain.go:67]: Okay--that makes sense.  So this is specifically sniffing out if the cert has been changed since the last time terraform has run.  

Noting that terraform should be used to manage cert deployment, it is worth asking what would happen if the cert was updated as part of the same plan.  I could make a strong argument that this is a "user error" edge case but we should cover our bases before submitting it to the larger community.  Is it even a valid concern?

File: internal/service/cognitoidp/user_pool_domain.go (hunk 7 above)
Comment (austinvalle) [reply internal/service/cognitoidp/user_pool_domain.go:177]: These were purposefully consistent with the destroy + create (consistently inconsistent lol), I believe the reasoning for why they are different is because of how the error messages are aggregated upstream, but that's a guess.

**Review by cmorton9249 — APPROVED** [review]

File: internal/service/cognitoidp/wait.go [hunk 9]
```diff
@@ -48,6 +48,31 @@ func waitUserPoolDomainCreated(conn *cognitoidentityprovider.CognitoIdentityProv
...
```
Comment (A078743) [inline internal/service/cognitoidp/wait.go:74]: I don't quite understand what's going on here. It seems like this returns a status of the domain status of the cognitoidentityprovider. Why would we return the timeout and not a message on timeout? Maybe this is just a convention that I'm not used to.

**Comment by A078743** [comment]
Reviewed to validate no brand terms or corporate specific entries were made. Deferring GO expertise to other reviewers. Thank you for making this change to bring the provider into line with the current API.

**Review by A078743 — APPROVED** [review]
One question comment left. Please review and make an update if appropriate.

File: internal/service/cognitoidp/wait.go (hunk 9 above)
Comment (austinvalle) [reply internal/service/cognitoidp/wait.go:74]: So line 74 is executing the config defined on line 63-72. It continually checks the status of the Cognito User Pool Domain until it enters the `Target` state, which is `Active`, or until it enters a state that is not `Pending`, which is `Updating`.

Once that condition is met, it will update state, and this function will return the result of the refreshed state.

If it times out in this operation, we don't need to further format the error, as this function `waitUserPoolDomainUpdated` does not handle the error, so we pass it along and let the higher up function handle the error.

Basically, there is no further formatting/messaging needed and we aren't handling the error here, so we pass it along.
